File: jsbridge/src/main/assets/WebViewJavascriptBridge.js
//notation: js file can only use this kind of comments
//since comments will cause error when use in webview.loadurl,
//comments will be remove by java use regexp
(function() {
    if (window.WebViewJavascriptBridge) {
        return;
    }

    // 进行url scheme传值的iframe
    var messagingIframe;
    // 发送消息队列
    var sendMessageQueue = [];
    // 接收消息队列
    var receiveMessageQueue = [];
    // 本地注册的方法集合，原生只能调用本地注册的方法，否则会提示错误
    var messageHandlers = {};

    // JsBridge协议定义的名称
    var JSBRIDGE_PROTOCOL_SCHEME = 'bridge://';
    var JSBRIDGE_RETURN_DATA = JSBRIDGE_PROTOCOL_SCHEME + 'return/';
    var JSBRIDGE_FETCH_QUEUE = JSBRIDGE_RETURN_DATA + '_fetchQueue/';
    var QUEUE_HAS_MESSAGE = '__QUEUE_MESSAGE__/';
    var UNDERLINE = '_';
    // 定义的回调函数集合，在原生调用完对应的方法后，会执行对应的回调函数id
    var responseCallbacks = {};
    // 唯一ID，用来确保每一个回调函数的唯一性
    var uniqueId = 1;

    function _createQueueReadyIframe(doc) {
        // 创建隐藏iframe过程
        messagingIframe = doc.createElement('iframe');
        messagingIframe.style.display = 'none';
        doc.documentElement.appendChild(messagingIframe);
    }

    // set default messageHandler
    function init(messageHandler) {
        if (WebViewJavascriptBridge._messageHandler) {
            throw new Error('WebViewJavascriptBridge.init called twice.');
        }
        WebViewJavascriptBridge._messageHandler = messageHandler;
        var receivedMessages = receiveMessageQueue;
        receiveMessageQueue = null;
        for (var i = 0; i < receivedMessages.length; i++) {
            _dispatchMessageFromNative(receivedMessages[i]);
        }
    }

    // 注册本地JS方法，注册后Native可通过JSBridge调用，调用后会将方法注册到本地变量messageHandlers中
    function registerHandler(handlerName, handler) {
        messageHandlers[handlerName] = handler;
    }

    // 调用原生开放的API，调用后实际上还是本地通过url scheme触发，调用时会将回调id存放到本地变量responseCallbacks中
    function callHandler(handlerName, data, responseCallback) {
        _doSend({
            handlerName: handlerName,
            data: data
        }, responseCallback);
    }

    function send(data, responseCallback) {
        _doSend({
            data: data
        }, responseCallback);
    }

    // sendMessage add message， 触发native处理 sendMessage
    function _doSend(message, responseCallback) {
        // 判断是否有回调函数，如果有则生成一个回调函数id，并将id和对应回调添加进入回调函数集合responseCallbacks中
        if (responseCallback) {
            var callbackId = 'callback_' + (uniqueId++) + UNDERLINE + new Date().getTime();
            responseCallbacks[callbackId] = responseCallback;
            message.callbackId = callbackId;
        }
        // 使用sendMessageQueue队列把消息存起来，并且改变Iframe的src，提醒native端来取消息
        sendMessageQueue.push(message);
        // 触发scheme
        messagingIframe.src = JSBRIDGE_PROTOCOL_SCHEME + QUEUE_HAS_MESSAGE;
    }

    // 提供给native调用，该函数作用:获取sendMessageQueue返回给native，由于android不能直接获取返回的内容，所以使用url shouldOverrideUrlLoading 的方式返回内容
    function _fetchQueue() {
        var messageQueueString = JSON.stringify(sendMessageQueue);
        sendMessageQueue = [];
        // android can't read directly the return data, so we can reload iframe src to communicate with java
        // 将js传递的参数拼接入Url中，传递到native，而相应的shouldOverrideUrlLoading()中调用handlerReturnData()这个方法
        // 注意：正常来说是可以通过window.location.href达到发起网络请求的效果的，但是有一个很严重的问题，就是如果连续多次修改window.location.href的值，在Native层只能接收到最后一次请求，前面的请求都会被忽略掉。所以JS端发起网络请求的时候，需要使用iframe，这样就可以避免这个问题。
        messagingIframe.src = JSBRIDGE_FETCH_QUEUE + encodeURIComponent(messageQueueString);
    }

    // 提供给native使用
    function _dispatchMessageFromNative(messageJSON) {
        setTimeout(function() {
            var message = JSON.parse(messageJSON);
            var responseCallback;
            // java call finished, now need to call js callback function
            // 假如存在responseId，则回调完成时删除responseId
            if (message.responseId) {
                responseCallback = responseCallbacks[message.responseId];
                if (!responseCallback) {
                    return;
                }
                responseCallback(message.responseData);
                delete responseCallbacks[message.responseId];
            } else {
                // 直接发送
                if (message.callbackId) {
                    var callbackResponseId = message.callbackId;
                    responseCallback = function(responseData) {
                        // js回调给java
                        _doSend({
                            responseId: callbackResponseId,
                            responseData: responseData
                        });
                    };
                }
                var handler = WebViewJavascriptBridge._messageHandler;
                if (message.handlerName) {
                    // 通过队列拿到handler
                    handler = messageHandlers[message.handlerName];
                }
                // 查找指定handler
                try {
                    handler(message.data, responseCallback);
                } catch (exception) {
                    if (typeof console != 'undefined') {
                        console.log("WebViewJavascriptBridge: WARNING: javascript handler threw.", message, exception);
                    }
                }
            }
        });
    }

    // 提供给native调用，receiveMessageQueue在会在页面加载完后赋值为null
    function _handleMessageFromNative(messageJSON) {
        console.log(messageJSON);
        if (receiveMessageQueue) {
            receiveMessageQueue.push(messageJSON);
        } else {
            // 处理来自Java层的主动调用
            _dispatchMessageFromNative(messageJSON);
        }
    }

    var WebViewJavascriptBridge = window.WebViewJavascriptBridge = {
        init: init,
        send: send,
        registerHandler: registerHandler,
        callHandler: callHandler,
        _fetchQueue: _fetchQueue,
        _handleMessageFromNative: _handleMessageFromNative
    };

    var doc = document;
    _createQueueReadyIframe(doc);
    var readyEvent = doc.createEvent('Events');
    readyEvent.initEvent('WebViewJavascriptBridgeReady');
    readyEvent.bridge = WebViewJavascriptBridge;
    doc.dispatchEvent(readyEvent);
})();

File: jsbridge/src/main/java/com/henley/jsbridge/browse/WebViewJavascriptBridge.java
package com.henley.jsbridge.browse;

import java.util.Collection;
import java.util.List;
import java.util.Map;

/**
 * BridgeWebView发消息的接口(定义了发消息的方法)
 */
interface WebViewJavascriptBridge {

    /**
     * 返回启动消息集合
     */
    List<Message> getStartupMessages();

    /**
     * 设置启动消息集合(启动消息处理完毕后会置为null)
     */
    void setStartupMessage(List<Message> startupMessages);

    /**
     * 设置默认处理程序
     *
     * @param handler 默认处理程序，用于处理由JavaScript发送的没有指定处理程序名称的消息(如果由JavaScript发送的消息指定了处理程序名称，则将由本地注册命名的处理程序处理)
     */
    void setDefaultHandler(BridgeHandler handler);

    void loadUrl(String jsUrl, Callback returnCallback);

    void loadUrl(String jsUrl, Map<String, String> additionalHttpHeaders, Callback returnCallback);

    /**
     * 批量注册处理程序以供JavaScript调用并响应JavaScript发送的消息
     *
     * @param handlerNames 处理程序名称集合
     * @param handler      处理程序(用于响应由JavaScript发送的指定处理程序名称的消息)
     */
    void registerHandler(Collection<String> handlerNames, JsHandler handler);

    /**
     * 注册处理程序以供JavaScript调用并响应JavaScript发送的消息
     *
     * @param handlerName 处理程序名称
     * @param handler     处理程序(用于响应由JavaScript发送的指定处理程序名称的消息)
     */
    void registerHandler(final String handlerName, final JsHandler handler);

    /**
     * 注册处理程序以供JavaScript调用并响应JavaScript发送的消息
     *
     * @param handlerNames 处理程序名称集合
     * @param handler      处理程序(用于处理由JavaScript发送的没有指定处理程序名称的消息)
     */
    void registerHandler(Collection<String> handlerNames, BridgeHandler handler);

    /**
     * 注册处理程序以供JavaScript调用并响应JavaScript发送的消息
     *
     * @param handlerName 处理程序名称
     * @param handler     处理程序(用于处理由JavaScript发送的没有指定处理程序名称的消息)
     */
    void registerHandler(String handlerName, BridgeHandler handler);

    /**
     * 批量调用JavaScript注册的处理程序
     *
     * @param handlerInfos 方法名称与参数的Map(方法名称为key，参数为value)
     * @param handler      处理程序(用于处理由JavaScript响应的消息)
     */
    void callHandler(Map<String, String> handlerInfos, JavaCallHandler handler);

    /**
     * 调用JavaScript注册的处理程序
     *
     * @param handlerName 处理程序名称
     * @param javaData    Native端传递给JS端的参数(JSON字符串)
     * @param handler     处理程序(用于处理由JavaScript响应的消息)
     */
    void callHandler(final String handlerName, String javaData, final JavaCallHandler handler);

    /**
     * 批量调用JavaScript注册的处理程序
     *
     * @param handlerInfos 方法名称与参数的Map(处理程序名称为Key，参数为Value)
     * @param callback     回调接口(用于处理由JavaScript响应的消息)
     */
    void callHandler(Map<String, String> handlerInfos, Callback callback);

    /**
     * 调用JavaScript注册的处理程序
     *
     * @param handlerName 方法名称
     * @param data        Native端传递给JS端的参数(JSON字符串)
     * @param callback    回调接口(用于处理由JavaScript响应的消息)
     */
    void callHandler(String handlerName, String data, Callback callback);

    /**
     * 向JavaScript发送消息
     *
     * @param data 消息内容
     */
    void send(String data);

    /**
     * 向JavaScript发送消息
     *
     * @param data     消息内容
     * @param callback 回调方法(用于处理由JavaScript响应的消息)
     */
    void send(String data, Callback callback);
}
